### PR TITLE
use oc instead of kubectl for OpenShift artefacts

### LINKED ISSUE #341: use oc insetad of kubectl for OpenShift artifacts
It is not possible to create OpenShift specific objects with `kubectl`

```
▶ ./kedge apply -f test.yaml
service "nginx" configured
error: error validating "STDIN": error validating data: couldn't find type: v1.DeploymentConfig; if you choose to ignore these errors, turn validation off with --validate=false
kubectl error: failed to execute command: exit status 1
```

For OpenShift artifacts `oc` has to be used instead of `kubectl` when calling `kedge apply`

### PR DESCRIPTION
**Pull request by kadel**
fixes #341
If OpenShift controller like DeploymentConfig is used in Kedge file,
than `oc` binary will be used instead of `kubectl` for commands like
`kedge apply`, `kedge create`, `kedge delete`

### CODE REVIEW

**Comment by kadel** [comment]
Don't understand why is Coveralls still complaining.
I've even added more tests than there was before.

File: pkg/cmd/commands.go [hunk 1]
```diff
@@ -50,16 +49,30 @@ func CreateKubernetesArtifacts(paths []string, generate bool, args ...string) er
...
 		if err != nil {
 			return errors.Wrap(err, "unable to perform controller operations")
 		}
 
+		// Get controller that is being used.
+		// This decides if whatever oc or kubectl will be used.
+		useOC := false
+		controller, err := spec.GetController(kedgeData)
+		if err != nil {
+			return errors.Wrap(err, "can't get controller")
```
Comment (containscafeine) [inline pkg/cmd/commands.go]: stupid nit: reword to "unable to get controller"

File: pkg/cmd/commands.go [hunk 2]
```diff
@@ -118,8 +131,14 @@ func CreateKubernetesArtifacts(paths []string, generate bool, args ...string) er
 	return nil
 }
 
-func runKubectl(args []string, data []byte) error {
-	cmd := exec.Command("kubectl", args...)
+// runClusterCommand calls kubectl or oc binary.
+// Booloan flag useOC controls if oc or kubectl will be used
```
Comment (containscafeine) [inline pkg/cmd/commands.go]: type for "boolean"

**Comment by containscafeine** [comment]
@kadel for now this is good, but then this will be using the `oc` binary only in case of DeploymentConfig, for everything else it'll be using `kubectl` - which will work fine - but then the end user needs to have both the binaries to deploy to OpenShift.
I think this is another place where we could use information from the `--distribution` flag.

This problem will still be there once we move to `kubernetes/client-go` and `openshift/client-go`.
`kubernetes/client-go` should still work with OpenShift, but it's just the inconsistency this brings to our operations since we have no information about the distribution this is being run on.

**Comment by kadel** [comment]
> @kadel for now this is good, but then this will be using the oc binary only in case of DeploymentConfig, for everything else it'll be using kubectl - which will work fine - but then the end user needs to have both the binaries to deploy to OpenShift.

Once we have Routes in the logic should change a bit. `oc` should be used if at least one OpenShift specific resource is used.



Once we stop using binaries, and use `openshift/client-go` this problem will go away.
It will be possible to use openshift library for everything, even for creating Kubernetes artifacts on Kubernetes

**Comment by kadel** [comment]
> I think this is another place where we could use information from the --distribution flag.

If stick with using `oc` and `kubectl` binaries, that it might make sense to add some flag like `--use-oc` to force calling `oc` for everything. I think that this might be better than `--distribution` flag.

**Comment by kadel** [comment]
Lets keep this open, until we merge Routes, than we can figure out how we will do this

**Comment by kadel** [comment]
I've updated this. 
Now it is not based on just a controller.

`oc` is used for all artifacts if there is at least one generated OpenShift  specific object (DeploymentConfig, Route) 

So if k8s Deployment is used as controller object, and there is a Route, than `oc` will be used.

**Comment by surajssd** [comment]
Rest LGTM!

**Comment by kadel** [comment]
> @kadel this should say oc error or something else, right now it says kubectl error.

replaced with "failed to execute command: ....."

**Comment by containscafeine** [comment]
@kadel  Haha, I hope I convince you for `--distribution` one day :laughing: 
LGTM!

**Pull request merged**